engine/table/src/main/java/io/deephaven/engine/table/impl/select/DynamicWhereFilter.java

```diff
@@ -241,79 +264,56 @@ public WritableRowSet filter(
-        }
-    }
-
-    private WritableRowSet filterLinearOne(RowSet selection, ColumnSource<?> keyColumn) {
+    private WritableRowSet filterLinear(RowSet selection, TupleSource<?> tupleSource) {
         if (selection.isEmpty()) {
             return RowSetFactory.empty();
         }
 
         if (!kernelValid) {
-            setInclusionKernel = SetInclusionKernel.makeKernel(keyColumn.getChunkType(), liveValues, inclusion);
+            setInclusionKernel = SetInclusionKernel.makeKernel(tupleSource.getChunkType(), liveValues, inclusion);
             kernelValid = true;
         }
 
         final RowSetBuilderSequential indexBuilder = RowSetFactory.builderSequential();
 
-        try (final ColumnSource.GetContext getContext = keyColumn.makeGetContext(CHUNK_SIZE);
-                final RowSequence.Iterator rsIt = selection.getRowSequenceIterator()) {
-            final WritableLongChunk<OrderedRowKeys> keyIndices = WritableLongChunk.makeWritableChunk(CHUNK_SIZE);
-            final WritableBooleanChunk<Values> matches = WritableBooleanChunk.makeWritableChunk(CHUNK_SIZE);
+        final int maxChunkSize = (int) Math.min(CHUNK_SIZE, selection.size());
+        // @formatter:off
+        try (final ColumnSource.GetContext keyGetContext = tupleSource.makeGetContext(maxChunkSize);
+             final RowSequence.Iterator selectionIterator = selection.getRowSequenceIterator();
+             final WritableLongChunk<OrderedRowKeys> selectionRowKeyChunk =
+                     WritableLongChunk.makeWritableChunk(maxChunkSize);
+             final WritableBooleanChunk<Values> matches = WritableBooleanChunk.makeWritableChunk(maxChunkSize)) {
+            // @formatter:on
 
-            while (rsIt.hasMore()) {
-                final RowSequence chunkOk = rsIt.getNextRowSequenceWithLength(CHUNK_SIZE);
+            while (selectionIterator.hasMore()) {
+                final RowSequence selectionChunk = selectionIterator.getNextRowSequenceWithLength(maxChunkSize);
 
-                final Chunk<Values> chunk = Chunk.downcast(keyColumn.getChunk(getContext, chunkOk));
-                setInclusionKernel.matchValues(chunk, matches);
+                final Chunk<Values> keyChunk = Chunk.downcast(tupleSource.getChunk(keyGetContext, selectionChunk));
+                final int thisChunkSize = keyChunk.size();
+                setInclusionKernel.matchValues(keyChunk, matches);
 
-                keyIndices.setSize(chunk.size());
-                chunkOk.fillRowKeyChunk(keyIndices);
+                selectionRowKeyChunk.setSize(thisChunkSize);
+                selectionChunk.fillRowKeyChunk(selectionRowKeyChunk);
 
-                for (int ii = 0; ii < chunk.size(); ++ii) {
+                for (int ii = 0; ii < thisChunkSize; ++ii) {
                     if (matches.get(ii)) {
-                        indexBuilder.appendKey(keyIndices.get(ii));
+                        indexBuilder.appendKey(selectionRowKeyChunk.get(ii));
```

Comment:
Is there value in trying to use `appendRange` instead? I know we've got arbitrary filtering here, but it's possible there is underlying structure and so we have large contiguous inclusions?

More generally, it seems like a utility wrapper around RowSetBuilderSequential could be generally useful in this regards?

Reply:
No. RowSetBuilderSequential should be optimizing range coalescing internally.